lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -169,6 +169,35 @@ static Value createSoftmaxBackwardCommonKernel(PatternRewriter &rewriter,
   return sub;
 }
 
+// This algorithm is adopted from:
+// https://github.com/pytorch/pytorch/blob/c47bdd7522c7146c091c95ff5b4ce354d7a8870c/aten/src/ATen/native/TensorShape.cpp#L3660
+static SmallVector<int64_t>
+computeDimsOrderForMoveDim(ArrayRef<int64_t> srcDimInt,
+                           ArrayRef<int64_t> dstDimInt, unsigned inputRank) {
+  SmallVector<int64_t> srcDims, dstDims, dimsOrder;
+  for (unsigned i = 0; i < inputRank; i++) {
+    srcDims.push_back(i);
+    dstDims.push_back(i);
+    dimsOrder.push_back(-1);
+  }
+  for (unsigned i = 0; i < srcDimInt.size(); i++) {
+    dimsOrder[dstDimInt[i]] = srcDimInt[i];
+    srcDims[srcDimInt[i]] = -1;
+    dstDims[dstDimInt[i]] = -1;
+  }
+  SmallVector<int64_t> sourceDims, destinationDims;
+  for (unsigned i = 0; i < inputRank; i++) {
+    if (srcDims[i] != -1)
+      sourceDims.push_back(srcDims[i]);
+    if (dstDims[i] != -1)
+      destinationDims.push_back(dstDims[i]);
+  }
+  unsigned restDim = inputRank - srcDimInt.size();
+  for (unsigned i = 0; i < restDim; i++)
+    dimsOrder[destinationDims[i]] = sourceDims[i];
+  return dimsOrder;
```

Comment:
Copying my comment over:

> I think you can do instead

> ```c++
>   SmallVector<int> dimsOrder(llvm::iota_range<int>(0, inputRank, /*inclusive=*/false));
>   dimsOrder.erase(dimsOrder.begin() + srcDimInt);
>   dimsOrder.insert(dimsOrder.begin() + dstDimInt, srcDimInt);
> ```

It might make it clearer to explain by example. Suppose you have

```python
input = torch.rand((2,3,4,5))
torch.movedim(input, 3, 1)
```

The result will be a tensor of shape `(2, 5, 3, 4)`. This is the same as doing

```python
torch.permute(input, (0, 3, 1, 2))
```

In other words, move the last dimension to the second dimension slot and shift all the other dimensions. This can be achieved by `pop`ing the last element of `(0, 1, 2, 3)` and inserting it in the index position `1`. This is what the C++ code I wrote above is doing: remove the `src` dimension from the list and re-insert it in the right location.

Reply:
Agree. Introducing new op will increase the burden of backends

Reply:
Thanks for the suggested change!